### PR TITLE
Fix itype shadowed variable

### PR DESCRIPTION
**Pull request by bkentel**
The main issue: `it_tool::ammo` shadowed `itype::ammo` unintentionally, and `it_comest` didn't initialize some member variables in its constructor. I fixed that and patched up the fallout from the change. While in `itype.h`I also finished cleaning it up a bit and removing includes etc.
- Removed extraneous includes, added forward declarations, patched up the files affected by the dropped includes (`rng.h` mostly).
- Added explicit underlying types to two enums to allow forward declarations.
- Flipped two usages of a typedef into the actual type.
- Dropped superfluous `public` from a few inheritance statements.
- Use in class initializers in `itype`.
- Use a defaulted destructor.
- Fix uninitialized values in `it_comest` and add in class initializers.
- Add `override`.
- Remove extraneous virtual inheritance.
- Use `std::move`.

### CODE REVIEW

**Comment by BevapDin** [comment]
`ammo_name` seems a bit misleading, it's actually the id of an ammo type. Name indicates something that is displayed to the user but not used in any internal logic.

**Comment by bkentel** [comment]
How about `ammo_id`?

**Comment by BevapDin** [comment]
Sounds good. Btw. why all the reordering of the members? And if you add override, you also remove the virtual from those functions as it's redundant anyway.

**Comment by bkentel** [comment]
Just shifted the order for alignment / padding warnings. I probably didn't have to move around as much as I did though.

File: src/iuse.cpp [hunk 1]
```diff
@@ -3334,7 +3334,7 @@ int iuse::extra_battery(player *p, item *, bool, point)
 {
     int inventory_index = g->inv_for_filter( _("Modify what?"), []( const item & itm ) {
         it_tool *tl = dynamic_cast<it_tool *>(itm.type);
-        return tl != nullptr && tl->ammo == "battery";
+        return tl != nullptr && tl->ammo_name == "battery";
```
Comment (kevingranade) [inline src/iuse.cpp]: Looks like you missed a few.

Comment (bkentel) [inline src/iuse.cpp]: Looks like it :grimacing: I'll get it.

**Comment by kevingranade** [comment]
jenkins says you're missing one at item.cpp:1084

**Comment by bkentel** [comment]
Had to run after that last fixup -- I'll take care of it.

File: src/itype.h [hunk 2]
```diff
@@ -395,51 +394,51 @@ struct itype {
...
-    const item_category *category; // category pointer or NULL for automatic selection
+    const item_category *category = nullptr; // category pointer or NULL for automatic selection
 
-    std::string snippet_category;
+    nc_color color = c_white; // Color on the map (color.h)
+    char sym = '#';       // Symbol on the ma
+
+    bool explode_in_fire() const
+    {
+        return explosion_on_fire_data.power >= 0;
```
Comment (DavidKeaton) [inline src/itype.h]: if power = 0, it's not to explode correct?

Comment (bkentel) [inline src/itype.h]: No idea, didn't modify that code -- only moved it.

**Comment by bkentel** [comment]
Looks like something in he lua bindings framework is causing an issue?

**Comment by BevapDin** [comment]
You need apply the renaming of `ammo` in the "lua/class_definitions.lua" file as well. I usually forget this, too.

**Comment by bkentel** [comment]
Ah ha! That's where it is :bulb: Thank you @BevapDin.

File: src/itype.h (hunk 2 above)
Comment (DavidKeaton) [inline src/itype.h]: I wonder then, as a power of 0 doesn't seem like it should explode, but that is not within the confines of this PR, so nvm. :-)

Comment (narc0tiq) [inline src/itype.h]: Maybe it means a 0-size explosion, i.e. the item is destroyed and nothing
else happens. Seems a bit odd, but there you have it.

Comment (DavidKeaton) [inline src/itype.h]: Good point narc, never thought of it that way. Is there a seperate function to test that property though?

Comment (narc0tiq) [inline src/itype.h]: I haven't the slightest clue.

**Comment by DavidKeaton** [comment]
Fair enough! Thanks for the prompt replies friend! :-)

File: src/itype.h (hunk 2 above)
Comment (BevapDin) [inline src/itype.h]: The values in `explosion_data` are directly forwarded to `game::explosion`. `explosion_data` is a simple struct supposed to only hold the values for a later call to `explosion`, it should not include any behavior. Using negative values as indicator of the validness is acceptable as a negative power value is clearly not valid for the explosion function. Whether 0 is a valid value is up for discussion, but adding this condition to `explosion_data`/`itype` seems wrong. What if the explosion function allows power = 0 in the future?

Technically this *could* be change to give power the default value `INT_MIN`, and check for `power!=INT_MIN`, but why?

**Pull request merged**